compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Prefixing.hs

```diff
@@ -31,14 +31,37 @@ import DA.Cli.Damlc.Command.MultiIde.Types
 
 -- ProgressToken Prefixing
 
--- We need to ensure all ProgressTokens from different subIDEs are unique to the client, so we prefix them.
--- We do not need to undo this transformation, as we'll store it and backwards lookup later.
+-- Progress tokens are created in 2 ways:
+--   The subIDE sends a CreateToken
+--     In this case, we need to prefix the token, as multiple subIDEs might use the same name 
+--   The client sends any message with a progress token attached
+--     In this case, we can assume the token is unique to the client, and not modify it
+-- There are then 2 messages that can use these tokens:
+--   Progress message from the subIDE
+--     This includes the token id, but we don't know if said token id was created by the subIDE (so needs prefixing) or client
+--     Therefore, whenever we see a token created by the server, we store it in a var (along with the subIde it came from)
+--     Then when we see a progress message with a token, if the token is in the var, we know it needs the prefix, otherwise its unmodified
+--   Cancel message from the client
+--     This has a token id, but again we dont know where it came from, so if it needs its prefix removed
+--     We again lookup in the map, if we find it, we remove the prefix, and send it to the subIDE that created it
+--     If we don't find it, we know its unique to the client, and only exists on one of the SubIDEs
+--     We are then safe to broadcast the cancel message to all IDEs, as those that don't have it will ignore the message
+-- TODO: Verify what happens if the client creates a token and sends to SUbIDE1, then subIDE tries to create one with the same name
+--   This gets prefixed and submitted to client
+--   after this, the client tries to cancel the token it created for SubIDE1, which gets submitted to all SubIDEs
+--   will this cancel the one subIDE2 made - yes
+--   So, can the server and client create clashing token names? Likely :(
```

Comment:
:(

Reply:
I wrote a fix, but its rather confusing, so I looked into whether or not server and client can create clashing names - the answer is no, the VSCode extension only ever generates full UUIDs (with the dashes and all), whereas ghcide uses "Unique", which is just a counter.
It might be better from a readability perspective to drop the change and write this up in a comment, lmk what you think.

The gist of the change is that now all progress tokens created by the client are prefixed with `client`, and inserted into the prefix map backwards (such that the key is the prefixed token, and the value is unprefixed, to match the usage behaviour of this map of converting sub-ide tokens to client tokens)
Also wrapped the subide home filepath in Maybe to signify client created tokens.
Leads to some messy logic when handling Progress reports from the server, and now we need to lookup the name we found in 2 ways - as though it was created by the server or by the client.
A possible refactor here would be to make the prefixes reversible, and only store the subide home in the mapping for server created tokens. I might try this tomorrow, but thinking about these prefixes/clashes hurts my brain, so we'll see :)